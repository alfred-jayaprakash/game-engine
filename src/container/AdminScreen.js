import React, {useState} from 'react';
import axios from 'axios';
import {
  Container,
  Row,
  Col,
  Alert,
  Button,
  Form,
  FormGroup,
  Input,
  Label,
} from 'reactstrap';
import AdminGamePanel from '../components/AdminGamePanel';

const SERVER_URL = process.env.REACT_APP_SERVER_URL || '/';

const AdminScreen = props => {
  const [room, setRoom] = useState ('');
  const [error, setError] = useState ('');
  const [category, setCategory] = useState ('office');
  const [time, setTime] = useState ('30');
  const [questions, setQuestions] = useState ('5');

  const onCreateRoom = e => {
    console.log ('URL to connect is ', SERVER_URL);
    axios
      .post (SERVER_URL + '/rooms', {
        room: 'New Room',
        config: {
          time,
          questions,
          category,
        },
      })
      .then (res => {
        if (res.data) {
          setRoom (res.data);
        }
      });
  };

  return (
    <Container
      className="d-flex flex-column justify-content-center bg-dark text-white vertical-center"
      fluid={true}
    >
      {error !== '' &&
        <Row>
          <Col>
            <Alert color="danger">{error}</Alert>
          </Col>
        </Row>}
      {room === '' &&
        <Form>
          <Row>
            <Col>
              <FormGroup>
                <Label for="category">Choose picture category:</Label>
                <Input
                  type="select"
                  className="form-control-lg"
                  name="selectcategory"
                  id="category"
                  onChange={e => setCategory (e.target.value)}
                >
                  <option selected={true} value="gn-office">Office</option>
                  <option value="gn-home">Home</option>
                  <option value="gn-nature">Nature</option>
                  <option value="gn-christian">Christian</option>
                  <option value="gn-modern">Modern</option>
                </Input>
              </FormGroup>
            </Col>
          </Row>
          <Row>
            <Col>
              <FormGroup>
                <Label for="count">Number of pictures:</Label>
                <Input
                  type="select"
                  name="selectcount"
                  className="form-control-lg"
                  id="count"
                  onChange={e => setQuestions (e.target.value)}
                >
                  <option value="3">3</option>
                  <option selected={true} value="5">5</option>
                  <option value="10">10</option>
                </Input>
              </FormGroup>
            </Col>
          </Row>
          <Row>
            <Col>
              <FormGroup>
                <Label for="duration">Time for each picture:</Label>
                <Input
                  type="select"
                  name="selectduration"
                  className="form-control-lg"
                  id="duration"
                  onChange={e => setTime (e.target.value)}
                >
                  <option value="20">20</option>
                  <option selected={true} value="30">30</option>
                  <option value="60">60</option>
                </Input>
              </FormGroup>
            </Col>
          </Row>
          <Row>
            <Col>
              <FormGroup>
                <Button
                  color="info"
                  className="btn-lg btn-block"
                  onClick={onCreateRoom}
                >
                  Create Room
                </Button>
              </FormGroup>
            </Col>
          </Row>
        </Form>}
      {room &&
        <Row>
          <Col>
            <AdminGamePanel
              room={room.id}
              error={error => setError (error)}
              server={SERVER_URL}
            />
          </Col>
        </Row>}
    </Container>
  );
};

export default AdminScreen;
