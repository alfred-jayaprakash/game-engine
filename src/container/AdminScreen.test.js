import React from 'react';
import {render, fireEvent, screen, act, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {shallow} from 'enzyme';
import AdminScreen from './AdminScreen';

let postFnMock;
let container;
let postPromise;
let axios;

describe ('Snapshot tests', () => {
  it ('should render correctly without issues', () => {
    const component = shallow (<AdminScreen />);

    expect (component).toMatchSnapshot ();
  });
});

describe ('Functional tests', () => {
  beforeAll (() => {
    //ARRANGE
    // tell jest not to mock gameroom
    jest.unmock ('axios');

    // require the actual module so that we can mock exports on the module
    axios = require.requireActual ('axios');
    let mockResponse = {
      data: {
        id: 121212,
      },
    };
    postPromise = Promise.resolve (mockResponse);
    postFnMock = jest.fn ((url, options) => postPromise);
    axios.post = postFnMock;
  });

  beforeEach (() => {
    container = document.createElement ('div');
    document.body.appendChild (container);
    postFnMock.mockClear ();
  });

  afterEach (() => {
    document.body.removeChild (container);
    container = null;
  });

  it ('initial screen should load correctly', () => {
    render (<AdminScreen />, container);
    //Make sure Pictopia options are shown
    expect (
      screen.getByLabelText ('Choose picture category:')
    ).toBeInTheDocument ();
    expect (screen.getByLabelText ('Number of pictures:')).toBeInTheDocument ();
    expect (
      screen.getByLabelText ('Time for each picture:')
    ).toBeInTheDocument ();
    expect (screen.getByText ('Create Room')).toBeInTheDocument ();
    //Make sure AdminGamePanel is not showing
    expect (screen.queryByText ('Start Game')).not.toBeInTheDocument ();
  });

  it ('should create room without changing options', async () => {
    //ACT
    render (<AdminScreen />, container);
    fireEvent.click (screen.getByText ('Create Room'));
    await act (() => postPromise); //Wait until AXIOS POST promise is resolved

    //ASSERT
    expect (postFnMock).toHaveBeenCalled (); //Submit to server should have happened once
    //Check the data that is passed to the server
    expect (postFnMock.mock.calls[0][1]).toEqual ({
      room: 'New Room',
      config: {
        time: '30',
        questions: '5',
        category: '11328847',
      },
    });
    //Make sure AdminGamePanel is showing
    expect (screen.getByText ('Start Game')).toBeInTheDocument ();
  });

  it ('should create room by changing game options', async () => {
    //ACT
    render (<AdminScreen />, container);
    userEvent.selectOptions (
      screen.getByLabelText ('Choose picture category:'),
      ['10757469']
    );
    userEvent.selectOptions (screen.getByLabelText ('Number of pictures:'), [
      '3',
    ]);
    userEvent.selectOptions (screen.getByLabelText ('Time for each picture:'), [
      '60',
    ]);
    fireEvent.click (screen.getByText ('Create Room'));
    await act (() => postPromise); //Wait until AXIOS POST promise is resolved

    //ASSERT
    expect (postFnMock).toHaveBeenCalled (); //Submit to server should have happened once
    //Check the data that is passed to the server
    expect (postFnMock.mock.calls[0][1]).toEqual ({
      room: 'New Room',
      config: {
        time: '60',
        questions: '3',
        category: '10757469',
      },
    });
    //Make sure AdminGamePanel is showing
    expect (screen.getByText ('Start Game')).toBeInTheDocument ();
  });

  it ('should show error when server fails', async () => {
    //ARRANGE
    let error = 'Server encountered error';
    let errorResponse = {
      data: {
        error,
      },
    };
    postPromise = Promise.resolve (errorResponse);
    postFnMock = jest.fn ((url, options) => postPromise);
    axios.post = postFnMock;

    //ACT
    render (<AdminScreen />, container);
    fireEvent.click (screen.getByText ('Create Room'));
    await act (() => postPromise); //Wait until AXIOS POST promise is resolved

    //Make sure error is displayed in screen
    expect (screen.getByText (error)).toBeInTheDocument ();

    //Make sure AdminGamePanel is not showing
    expect (screen.queryByText ('Start Game')).not.toBeInTheDocument ();
  });
});
