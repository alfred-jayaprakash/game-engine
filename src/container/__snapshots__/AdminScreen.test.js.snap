// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests should render correctly without issues 1`] = `
<Container
  className="d-flex flex-column justify-content-center bg-dark text-white vertical-center"
  fluid={true}
  tag="div"
>
  <Form
    tag="form"
  >
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <FormGroup
          tag="div"
        >
          <Label
            for="category"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Choose picture category:
          </Label>
          <Input
            className="form-control-lg"
            id="category"
            name="selectcategory"
            onChange={[Function]}
            type="select"
            value="gn-office"
          >
            <option
              value="11328847"
            >
              Random
            </option>
            <option
              value="10757469"
            >
              Office
            </option>
            <option
              value="11328818"
            >
              Family
            </option>
            <option
              value="11328826"
            >
              Nature
            </option>
            <option
              value="11328790"
            >
              Christian
            </option>
          </Input>
        </FormGroup>
      </Col>
    </Row>
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <FormGroup
          tag="div"
        >
          <Label
            for="count"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Number of pictures:
          </Label>
          <Input
            className="form-control-lg"
            id="count"
            name="selectcount"
            onChange={[Function]}
            type="select"
            value="5"
          >
            <option
              value="3"
            >
              3
            </option>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="10"
            >
              10
            </option>
          </Input>
        </FormGroup>
      </Col>
    </Row>
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <FormGroup
          tag="div"
        >
          <Label
            for="duration"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Time for each picture:
          </Label>
          <Input
            className="form-control-lg"
            id="duration"
            name="selectduration"
            onChange={[Function]}
            type="select"
            value="30"
          >
            <option
              value="20"
            >
              20
            </option>
            <option
              value="30"
            >
              30
            </option>
            <option
              value="60"
            >
              60
            </option>
          </Input>
        </FormGroup>
      </Col>
    </Row>
    <Row
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <FormGroup
          tag="div"
        >
          <Button
            className="btn-lg btn-block"
            color="info"
            onClick={[Function]}
            tag="button"
          >
            Create Room
          </Button>
        </FormGroup>
      </Col>
    </Row>
  </Form>
</Container>
`;
